### PR TITLE
[DRAFT] Move accounts vuex state to composable

### PR DESCRIPTION
**Pull request by peronczyk**

### CODE REVIEW

File: src/protocols/aeternity/helpers/index.ts [hunk 1]
```diff
@@ -124,6 +125,17 @@ export function categorizeContractCallTxObject(transaction: ITransaction): {
...
 }
 
+export function convertMultisigAccountToAccount(
+  multisigAccount: IMultisigAccount,
+): Partial<IAccount> {
+  return {
+    address: multisigAccount.gaAccountId,
+    idx: 0,
+    protocol: PROTOCOL_AETERNITY,
+    globalIdx: 0,
```
Comment (pstachel) [inline src/protocols/aeternity/helpers/index.ts:135]: Why will they always have idx equal to 0?

File: src/protocols/aeternity/libs/AeSdkSuperhero.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable class-methods-use-this */
```
Comment (pstachel) [inline src/protocols/aeternity/libs/AeSdkSuperhero.ts:1]: is it needed?

File: src/protocols/bitcoin/libs/BitcoinAdapter.ts [hunk 3]
```diff
@@ -153,6 +153,7 @@ export class BitcoinAdapter extends BaseProtocolAdapter {
     };
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (pstachel) [inline src/protocols/bitcoin/libs/BitcoinAdapter.ts:156]: Isn't it used?

File: src/composables/accounts.ts [hunk 4]
```diff
@@ -122,16 +149,11 @@ export function useAccounts({ store }: IDefaultComposableOptions) {
...
-    } else {
-      protocolNextAccountIdx.value[protocol]! += 1;
-    }
-  }
-
-  function setProtocolNextAccountIdx(value: number, protocol: Protocol) {
-    protocolNextAccountIdx.value[protocol] = value;
+  function getLastProtocolAccount(protocol: Protocol): IAccount | undefined {
+    const protocolAccounts = accountsGroupedByProtocol.value[protocol];
+    return protocolAccounts
```
Comment (pstachel) [inline src/composables/accounts.ts:154]:
```suggestion
    return (protocolAccounts)
```

File: src/protocols/bitcoin/libs/BitcoinAdapter.ts (hunk 3 above)
Comment (peronczyk) [reply src/protocols/bitcoin/libs/BitcoinAdapter.ts:156]: True. Leftover that needs to be removed.

File: src/protocols/aeternity/libs/AeSdkSuperhero.ts (hunk 2 above)
Comment (peronczyk) [reply src/protocols/aeternity/libs/AeSdkSuperhero.ts:1]: Unfortunately yes :/ The changes done in the `getAccounts` method causes the error as there is no `this` reference now.

File: src/protocols/aeternity/helpers/index.ts (hunk 1 above)
Comment (peronczyk) [reply src/protocols/aeternity/helpers/index.ts:135]: Good point. The reason is that we don't use indexes when displaying the multisig accounts. For example here are is the card:
image
I'll add a comment about the reason why both idx and globalIdx are equal to 0.
What do you think?

File: src/popup/components/TransactionLabel.vue [hunk 5]
```diff
@@ -135,6 +135,8 @@ export default defineComponent({
 
     const availableTokens = useState<ITokenList>('fungibleTokens', 'availableTokens');
 
+    const getDefaultName = store.getters['names/getDefault'] as (a?: string) => string | undefined;
```
Comment (konradodo) [inline src/popup/components/TransactionLabel.vue:138]: i think, this getter should be moved to utils function in this pr

File: src/popup/components/NameItem.vue [hunk 6]
```diff
@@ -200,8 +200,10 @@ export default defineComponent({
     const error = ref(false);
     const pointerInput = ref();
 
+    const getDefaultName = store.getters['names/getDefault'] as (a?: string) => string | undefined;
```
Comment (konradodo) [inline src/popup/components/NameItem.vue:203]: i think, this getter should be moved to utils function in this pr

**Review by konradodo — CHANGES_REQUESTED** [review]

File: src/composables/accounts.ts [hunk 7]
```diff
@@ -1,94 +1,122 @@
...
+const accounts = computed((): IAccount[] => {
+  if (!mnemonic.value || !accountsRaw.value.length) {
+    return [];
+  }
+
+  return accountsRaw.value
+    .map((account, index) => {
+      const hdWallet = ProtocolAdapterFactory
+        .getAdapter(account.protocol)
+        .getHdWalletAccountFromMnemonicSeed(mnemonicSeed.value, account.idx) as any;
```
[the remark this answers is not in the document]
Comment (peronczyk) [reply src/composables/accounts.ts:64]: Totally right. The `any` here is only to satisfy the difference in the `IAccount.address` type. Currently we use `Encoded.AccountAddress` but the HDWallet returns `string`. We will update this soon. I'll add `TODO` comment here. OK?

File: src/popup/components/NameItem.vue (hunk 6 above)
Comment (peronczyk) [reply src/popup/components/NameItem.vue:203]: This makes no sense as next week we will replace Vuex "names" module with composable. So this line is temporary.

**Review by konradodo — APPROVED** [review]

File: src/composables/accounts.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (pstachel) [reply src/composables/accounts.ts:64]: Of course, I'm resolving the comment then

File: src/protocols/aeternity/helpers/index.ts (hunk 1 above)
Comment (pstachel) [reply src/protocols/aeternity/helpers/index.ts]: I agree. It's always better to have an explanation

File: src/protocols/aeternity/libs/AeSdkSuperhero.ts (hunk 2 above)
Comment (pstachel) [reply src/protocols/aeternity/libs/AeSdkSuperhero.ts:1]: Ok, resolving then

File: src/composables/accounts.ts [hunk 8]
```diff
@@ -115,47 +169,74 @@ export function useAccounts({ store }: IDefaultComposableOptions) {
...
 
+  function setActiveAccountByProtocolAndIdx(protocol: Protocol, idx: number) {
+    const accountFound = accountsGroupedByProtocol.value[protocol]
+      ?.find((account) => account.idx === idx);
+    if (accountFound) {
+      setActiveAccountByGlobalIdx(accountFound.globalIdx);
+    }
+  }
+
+  function setMnemonic(val: string) {
```
Comment (pstachel) [inline src/composables/accounts.ts:180]: Maybe you could consider more meaningful name than `val`?

File: src/composables/accounts.ts [hunk 9]
```diff
@@ -115,47 +169,74 @@ export function useAccounts({ store }: IDefaultComposableOptions) {
...
+    protocol,
+  }: Omit<IAccountRaw, 'type'>): number {
+    accountsRaw.value.push({
+      protocol,
+      isRestored,
+      type: ACCOUNT_HD_WALLET,
+    });
+    const lastProtocolAccount = getLastProtocolAccount(protocol);
+    const idx: number = (lastProtocolAccount) ? lastProtocolAccount.idx : 0;
+    return idx;
```
Comment (pstachel) [inline src/composables/accounts.ts:206]: I think that assignment is not needed here
```suggestion
    return (lastProtocolAccount) ? lastProtocolAccount.idx : 0;
```

**Review by pstachel — APPROVED** [review]

File: src/store/plugins/fungibleTokens.js [hunk 10]
```diff
@@ -220,10 +219,11 @@ export default (store) => {
...
   );
 
-  store.watch(
-    () => aeNextAccountIdx.value,
+  // Fetch balances whenever new account is added
+  watch(
+    aeAccounts,
     async (val, oldVal) => {
-      if (val !== oldVal) {
+      if (val.length !== oldVal.length) {
```
Comment (subhod-i) [inline src/store/plugins/fungibleTokens.js:226]: Please test this with a fresh new wallet and also when restoring a wallet. `oldval` could be undefined and accessing `.length` may lead to runtime errors`

File: src/store/mutations.js [hunk 11]
```diff
@@ -10,9 +10,6 @@ export default {
   setChainNames(state, payload) {
     state.chainNames = payload;
   },
-  setMnemonic(state, payload) {
-    state.mnemonic = payload;
-  },
```
Comment (subhod-i) [inline src/store/mutations.js:15]: Where did we move this?

File: src/store/modules/accounts/index.js [hunk 12]
```diff
@@ -14,65 +8,16 @@ export default {
...
     },
 
-    signTransaction({ getters: { active, getModule }, dispatch }, { txBase64, options }) {
+    signTransaction({ dispatch }, { txBase64, options }) {
       if (options && options.fromAccount) {
-        return dispatch(`${getModule(active).name}/signTransactionFromAccount`, { txBase64, options });
+        return dispatch('hdWallet/signTransactionFromAccount', { txBase64, options });
       }
-      return dispatch(`${getModule(active).name}/signTransaction`, { txBase64, options });
+      return dispatch('hdWallet/signTransaction', { txBase64, options });
```
Comment (subhod-i) [inline src/store/modules/accounts/index.js:20]: We may support only `hdWallet` for now but we had the provisions in place to support ledger wallet. I am not sure why this PR is dropping the support of hardware ledger accounts.

File: src/protocols/aeternity/libs/AeSdkSuperhero.ts
Comment (subhod-i) [inline src/protocols/aeternity/libs/AeSdkSuperhero.ts:1]: Since the class property `store` is not being used, Please remove the property from class and it's constructor

```
  store: Store<any>; // This one
  constructor(store: Store<any>, options: any) {
    super(options);
    this.store = store;
  }
 
  ```

**Review by subhod-i — CHANGES_REQUESTED** [review]

File: src/composables/accounts.ts (hunk 8 above)
Comment (peronczyk) [reply src/composables/accounts.ts:180]: `mnemonic` is already used. Any suggestions?

File: src/utils/common.ts [hunk 13]
```diff
@@ -364,7 +374,10 @@ export function detectProtocolByOwner(network: string, address?: string) {
   return null;
 }
 
-export async function defaultAccountDiscovery(
+/**
+ * @returns {number} between -1 and n where -1 means there are no accounts found.
+ */
+export async function getLastNotEmptyAccountIndex(
```
[the remark this answers is not in the document]
Comment (peronczyk) [reply src/utils/common.ts:380]: I kind of understand what you want to achieve, but this requires really complex TS custom typings which in my opinion are an overkill. Especially as the TypeScript itself types and comments stuff the same way as I did. For example here's a typing info for `Array.findIndex()` which returns `-1...n`:
image
You see the return value is typed as `number` but there is an additional comment about the possible values.

File: src/store/plugins/fungibleTokens.js (hunk 10 above)
Comment (peronczyk) [reply src/store/plugins/fungibleTokens.js:226]: `aeAccunts` is based on regular `accounts` which always returns an array. If user didn't restored the accounts the array is empty.

File: src/protocols/aeternity/libs/AeSdkSuperhero.ts
Comment (peronczyk) [reply src/protocols/aeternity/libs/AeSdkSuperhero.ts:1]: I wanted to achieve this, but the method `_resolveAccount` returns the `AccountSuperhero` which also has the `store` as dependency :/ If you see how we can remove this please let me know - I would be so happy to remove it :)

File: src/store/modules/accounts/index.js (hunk 12 above)
Comment (peronczyk) [reply src/store/modules/accounts/index.js:20]: We didn't support the ledger. Please take a closer look on the PR and this line: https://github.com/aeternity/superhero-wallet/blob/develop/src/store/modules/accounts/index.js#L11

File: src/popup/components/Modals/AccountCreate.vue [hunk 14]
```diff
@@ -73,10 +73,11 @@ export default defineComponent({
           break;
 
         case PROTOCOL_BITCOIN:
-          await store.dispatch('accounts/hdWallet/create', {
+          idx = addRawAccount({
             isRestored: false,
             protocol: PROTOCOL_BITCOIN,
           });
+          setActiveAccountByProtocolAndIdx(PROTOCOL_BITCOIN, idx);
```
[the remark this answers is not in the document]
Comment (peronczyk) [reply src/popup/components/Modals/AccountCreate.vue:80]: No it's not used, but it's not possible to create `let`/`const` within the `case` block.

File: src/composables/accounts.ts (hunk 8 above)
Comment (pstachel) [reply src/composables/accounts.ts:180]: I'm thinking of `phrase` or `mnemonicPhrase` to indicate what kind of value will be passed here

File: src/composables/composablesHelpers.ts [hunk 15]
```diff
@@ -33,7 +33,7 @@ interface ICreateStorageRefOptions<T> {
  */
 export function useStorageRef<T = string | object | any[]>(
   initialState: T,
-  keys: StorageKeysInput,
+  keys: StorageKey,
```
Comment (subhod-i) [inline src/composables/composablesHelpers.ts:36]:
```suggestion
  key: StorageKey,
```

Not plural.

**Review by subhod-i — CHANGES_REQUESTED** [review]

File: src/composables/composablesHelpers.ts (hunk 15 above)
Comment (peronczyk) [reply src/composables/composablesHelpers.ts:36]: OK. Done.

File: src/utils/common.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (subhod-i) [reply src/utils/common.ts:380]: Is it really complex?

The typescript document itself uses `number` but IMO we can do better. I am pushing this because it's being used at  crucial helper methods and it's important to indicate this for someone to use this in the future.

File: src/store/modules/accounts/index.js (hunk 12 above)
Comment (subhod-i) [reply src/store/modules/accounts/index.js:20]: Sure. It's better to have a module-based routing but that can be introduced when adding ledger wallet feature too.

File: src/composables/transactionTx.ts [hunk 16]
```diff
@@ -188,7 +187,7 @@ export function useTransactionTx({
         return {
           name: getPreferredName.value(address) || '',
           address,
-        } as IAccount; // TODO establish the required return type for the function
+        } as any; // TODO establish the required return type for the function
```
[the remark this answers is not in the document]
Comment (peronczyk) [reply src/composables/transactionTx.ts:190]: `IAccount` was actually not correct. I'm changing this to `IAccountOverview`.

File: src/protocols/aeternity/libs/AeSdkSuperhero.ts
Comment (subhod-i) [reply src/protocols/aeternity/libs/AeSdkSuperhero.ts:1]: Oho, I didn't see the `_resolveAccount`. Something to tackle when we move the signing methods out of store.

File: src/utils/common.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (peronczyk) [reply src/utils/common.ts:380]: There are 2 issues:

1. There is no native range type. There are some ways to do it (https://stackoverflow.com/questions/39494689/is-it-possible-to-restrict-number-to-a-certain-range) but it only makes sense for small ranges. When you try to make a large range this is what you get:
image

2. There is no typing for the infinity. When you try to use `typeof Infinity` TS has no knowledge how to handle it and returns `never`. Take a look:
image

So in short: there is a reason why TS engine does not provide any range typings for native JS functions. Instead it uses `number`. And I don't know any programming language that makes this easy or even possible.

File: src/composables/accounts.ts (hunk 8 above)
Comment (peronczyk) [reply src/composables/accounts.ts:180]: I'll go with `newMnemonic` as agreed on private.

File: src/utils/common.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (subhod-i) [reply src/utils/common.ts:380]: https://github.com/microsoft/TypeScript/issues/54925
Yup, the issue is still open. We can resolve this.

**Review by subhod-i — APPROVED** [review]
Code changes LGTM, We need to test this if the changes in the PR can read the existing storage keys and their data. @Liubov-crypto Please use the develop branch and make sure to have an account with few interactions and then use this PR to check the existing state is continued/resumed.

File: src/popup/components/Modals/MultisigVaultCreate.vue [hunk 17]
```diff
@@ -26,8 +26,8 @@
           v-model.trim="signer.address"
           :default-text="$t('modals.createMultisigAccount.signerInputPlaceholder')"
           :label="getSignerLabel(index)"
-          item-title="address"
           :options="aeAccountsSelectOptions"
+          item-title="value"
```
Comment (CedrikNikita) [inline src/popup/components/Modals/MultisigVaultCreate.vue:30]: what is `value` in this context?

**Review by CedrikNikita — APPROVED** [review]
Code changes look reasonable.

File: src/popup/components/Modals/MultisigVaultCreate.vue (hunk 17 above)
Comment (peronczyk) [reply src/popup/components/Modals/MultisigVaultCreate.vue:30]: This is a common concept for custom Vue select components where the items in the list are objects. So you are able to define which object property (key) is used as the title and the value. Reference: https://vuetifyjs.com/en/api/v-select/#props-item-title
